crates/lair_keystore_api/src/encoding_types.rs

```diff
@@ -0,0 +1,257 @@
+        serializer.serialize_str(&s)
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for BinData {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let tmp: String = serde::Deserialize::deserialize(deserializer)?;
+        base64::decode_config(&tmp, base64::URL_SAFE_NO_PAD)
+            .map_err(serde::de::Error::custom)
+            .map(|b| Self(b.into()))
+    }
+}
+
+/// Wrapper newtype for serde encoding / decoding sized binary data
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct BinDataSized<const N: usize>(pub Arc<[u8; N]>);
+
+impl<const N: usize> std::fmt::Debug for BinDataSized<N> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let s = base64::encode_config(&*self.0, base64::URL_SAFE_NO_PAD);
+        write!(f, "BinDataSized<{}>({})", N, s)
+    }
+}
+
+impl<const N: usize> std::fmt::Display for BinDataSized<N> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let s = base64::encode_config(&*self.0, base64::URL_SAFE_NO_PAD);
+        f.write_str(&s)
+    }
+}
+
+impl<const N: usize> BinDataSized<N> {
+    /// Get a clone of our inner Arc<[u8; N]>
+    pub fn cloned_inner(&self) -> Arc<[u8; N]> {
+        self.0.clone()
+    }
+}
+
+impl<const N: usize> From<[u8; N]> for BinDataSized<N> {
+    fn from(b: [u8; N]) -> Self {
+        Self(Arc::new(b))
+    }
+}
+
+impl<const N: usize> From<Arc<[u8; N]>> for BinDataSized<N> {
+    fn from(b: Arc<[u8; N]>) -> Self {
+        Self(b)
+    }
+}
+
+impl<const N: usize> std::ops::Deref for BinDataSized<N> {
+    type Target = [u8; N];
+
+    fn deref(&self) -> &Self::Target {
+        &*self.0
+    }
+}
+
+impl<const N: usize> serde::Serialize for BinDataSized<N> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let s = base64::encode_config(&*self.0, base64::URL_SAFE_NO_PAD);
+        serializer.serialize_str(&s)
+    }
+}
+
+impl<'de, const N: usize> serde::Deserialize<'de> for BinDataSized<N> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let tmp: String = serde::Deserialize::deserialize(deserializer)?;
+        let tmp = base64::decode_config(&tmp, base64::URL_SAFE_NO_PAD)
+            .map_err(serde::de::Error::custom)?;
+        if tmp.len() != N {
+            return Err(serde::de::Error::custom("invalid buffer length"));
+        }
+        let mut out = [0; N];
+        out.copy_from_slice(&tmp);
+        Ok(Self(Arc::new(out)))
+    }
+}
+
+impl BinDataSized<32> {
+    /// Treat this bin data as an ed25519 public key,
+    /// and use it to verify a signature over a given message.
+    pub async fn verify_detached<M>(
+        &self,
+        signature: BinDataSized<64>,
+        message: M,
+    ) -> LairResult<bool>
```

Comment:
Not a big deal but seems you don't need the self

```suggestion
    pub async fn verify_detached<M>(
        signature: BinDataSized<64>,
        message: M,
    ) -> LairResult<bool>
```

Reply:
not sure I follow... how do you know what the public key is without `self`?